File: kotlin/protein-translation/src/main/kotlin/ProteinTranslation.kt
fun translate(rna: String?): List<String> {
    return rna?.let { rna ->
        rna
            .chunked(3)
            .map { translateSequence(it) }
            .takeWhile { it != "STOP" }
            .also {
                it.forEach {
                    require(it != "")
                }
            }
    } ?: emptyList()
}

fun translateSequence(rna: String): String = when (rna) {
    "AUG" -> "Methionine"
    "UUU", "UUC" -> "Phenylalanine"
    "UUA", "UUG" -> "Leucine"
    "UCU", "UCC", "UCA", "UCG" -> "Serine"
    "UAU", "UAC" -> "Tyrosine"
    "UGU", "UGC" -> "Cysteine"
    "UGG" -> "Tryptophan"
    "UAA", "UAG", "UGA" -> "STOP"
    else -> ""
}

File: csharp/protein-translation/ProteinTranslation.cs
using System;
using System.Collections.Generic;
using System.Linq;

using Microsoft.VisualBasic;

public static class ProteinTranslation
{
    private static readonly Dictionary<string, string[]> ProteinList = new()
    {
        {"Methionine", new[] {"AUG"}}, 
        {"Phenylalanine", new[] {"UUU", "UUC"}},
        {"Leucine", new[] {"UUA", "UUG"}},
        {"Serine", new[] {"UCU", "UCC", "UCA", "UCG"}},
        {"Tyrosine", new[] {"UAU", "UAC"}},
        {"Cysteine", new[] {"UGU", "UGC"}},
        {"Tryptophan", new[] {"UGG"}},
        {"STOP", new[] {"UAA", "UAG", "UGA"}},
    };

    public static string[] Proteins(string strand)
    {
        int length = strand.Length / 3;
        List<string> result = [];
        for (int i = 0; i < length; i++)
        {
            string prot = FindProtein(strand.Substring(i * 3, 3));

            if (prot == "STOP")
            {
                break;
            }
            
            result.Add(prot);
        }

        return [.. result];
    }

    private static string FindProtein(string strand) => ProteinList.FirstOrDefault(x => x.Value.Contains(strand)).Key;
}